Add flexmock, sqlalchemy-utils, pgpy, and py-gfm

bsima opened the pull request:
###### Motivation for this change

Trying to build [core.sr.ht](https://git.sr.ht/~sircmpwn/core.sr.ht) for NixOS and running into missing packages, figured I would add them upstream. I tested each one with `nix-build -A pythonPackages.<pkgname>` to make sure they built.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

bsima committed:
pythonPackages.flexmock: -> 0.10.2

bsima committed:
pythonPackages.sqlalchemy-utils: -> 0.33.6

bsima committed:
pythonPackages.pgpy: -> 0.4.3

bsima committed:
pythonPackages.py-gfm: -> 0.1.4

bsima committed:
pythonPackages.sshpubkeys: -> 3.1.0

lsix reviewed (changes requested):
Hi,

Thanks for the PR.

Could you add `meta` attributes for `flexmock`, `PGPy` and `py-gfm` ?

bsima committed:
pythonPackages.sqlalchemy-utils: fix license

bsima committed:
pythonPackages.py-gfm: add meta

bsima committed:
pythonPackages.pgpy: add meta

bsima committed:
pythonPackages.flexmock: add meta

bsima committed:
pythonPackages.sshpubkeys: add meta

bsima commented:
Lancelot SIX <notifications@github.com> writes:

> lsix requested changes on this pull request.
>
> Hi,
>
> Thanks for the PR.
>
> Could you add `meta` attributes for `flexmock`, `PGPy` and `py-gfm` ?

Okay, I added the `meta` attributes, and added one more package,
'sshpubkeys'. Let me know if you need any more changes :)

worldofpeace commented on pkgs/development/python-modules/py-gfm/default.nix:
```diff
@@ -0,0 +1,23 @@
+
+  meta = with lib; {
+    homepage = https://github.com/zopieux/py-gfm;
+    description = "This is an implementation of GitHub-Flavored Markdown written as an extension to the Python Markdown library. It aims for maximal compatibility with GitHub's rendering.";
```
This description is too long.

bsima committed:
pythonPackages.py-gfm: Shorten length of description

bsima commented:
worldofpeace <notifications@github.com> writes:

> +  meta = with lib; {
> +    homepage = https://github.com/zopieux/py-gfm;
> +    description = "This is an implementation of GitHub-Flavored Markdown written as an extension to the Python Markdown library. It aims for maximal compatibility with GitHub's rendering.";
>
> This description is too long.

Fixed in 9e0430cb919

dotlambda commented on pkgs/development/python-modules/flexmock/default.nix:
```diff
@@ -0,0 +1,24 @@
+    sha256 = "fe95c8727f4db73dc8f2f7b4548bffe7992440a965fefd60da291abda5352c2b";
+  };
+
+  buildInputs = [ nose pytest ];
```
```suggestion
  checkInputs = [ nose pytest ];
```

dotlambda commented on pkgs/development/python-modules/flexmock/default.nix:
```diff
@@ -0,0 +1,24 @@
+    homepage = https://flexmock.readthedocs.io/en/latest/;
+    description = "flexmock is a testing library for Python.";
+    license = licenses.bsd2;
+  };
```
`maintainers`?

dotlambda commented on pkgs/development/python-modules/pgpy/default.nix:
```diff
@@ -0,0 +1,26 @@
+    sha256 = "04412dddd6882ac0c5d5daf4326c28d481421851a68e25e7ac8e06cc9dc2b902";
+  };
+
+  buildInputs = [ cryptography pyasn1 six singledispatch ];
```
```suggestion
  propagatedBuildInputs = [ cryptography pyasn1 six singledispatch ];
```

dotlambda commented on pkgs/development/python-modules/pgpy/default.nix:
```diff
@@ -0,0 +1,26 @@
+
+  meta = with lib; {
+    homepage = https://github.com/SecurityInnovation/PGPy;
+    description = "PGPy is a Python (2 and 3) library for implementing Pretty Good Privacy into Python programs, conforming to the OpenPGP specification per RFC 4880.";
```
https://nixos.org/nixpkgs/manual/#sec-standard-meta-attributes

dotlambda commented on pkgs/development/python-modules/pgpy/default.nix:
```diff
@@ -0,0 +1,26 @@
+    homepage = https://github.com/SecurityInnovation/PGPy;
+    description = "PGPy is a Python (2 and 3) library for implementing Pretty Good Privacy into Python programs, conforming to the OpenPGP specification per RFC 4880.";
+    license = licenses.bsd3;
+  };
```
`maintainers`?

dotlambda commented on pkgs/development/python-modules/flexmock/default.nix:
```diff
@@ -0,0 +1,24 @@
+
+  meta = with lib; {
+    homepage = https://flexmock.readthedocs.io/en/latest/;
+    description = "flexmock is a testing library for Python.";
```
https://nixos.org/nixpkgs/manual/#sec-standard-meta-attributes

dotlambda reviewed (changes requested):
The requested changes should be done for the packages I didn't comment on as well.
You might want to read through https://nixos.org/nixpkgs/manual/#python again.

bsima committed:
pythonPackages.flexmock: Change 'buildInputs' to 'checkInputs'

bsima committed:
maintainers: Add bsima

bsima committed:
pythonPackages.flexmock: Add maintainer

bsima committed:
pythonPackages.pgpy: Change 'buildInputs' to 'propogatedBuildInputs'

bsima committed:
pythonPackages.pgpy: Split description into two attributes

bsima committed:
pythonPackages.pgpy: Add meta info

bsima committed:
pythonPackages.py-gfm: Add meta

bsima committed:
pythonPackages.sshpubkeys: Add meta

bsima committed:
pythonPackages.sqlalchemy-utils: Add meta

bsima commented:
Robert Schütz <notifications@github.com> writes:

> dotlambda requested changes on this pull request.
>
> The requested changes should be done for the packages I didn't comment on as well.

Okay, I think I covered all of them. They all have meta, metaintainers, etc.

> You might want to read through https://nixos.org/nixpkgs/manual/#python again.

Looking at the contributing guidelines, this is what I see:

> Python libraries are called from python-packages.nix and packaged with
> buildPythonPackage. The expression of a library should be in
> pkgs/development/python-modules/<name>/default.nix. Libraries in
> pkgs/top-level/python-packages.nix are sorted quasi-alphabetically to
> avoid merge conflicts.

Check.

> Python applications live outside of python-packages.nix and are packaged
> with buildPythonApplication.

None of these are apps.

> Make sure libraries build for all Python interpreters.

I think they all succeed. I'll look into testing this thoroughly over
the weekend or next week.

> By default we enable tests. Make sure the tests are found and, in the
> case of libraries, are passing for all interpreters. If certain tests
> fail they can be disabled individually. Try to avoid disabling the tests
> altogether. In any case, when you disable tests, leave a comment
> explaining why.

Only one package failed to find the tests, and I added a comment
explaining that.

> Commit names of Python libraries should reflect that they are Python
> libraries, so write for example pythonPackages.numpy: 1.11 -> 1.12.

Check.

> Attribute names in python-packages.nix should be normalized according to
> PEP 0503. This means that characters should be converted to lowercase
> and . and _ should be replaced by a single - (foo-bar-baz instead of
> Foo__Bar.baz )

Check.

Okay, anything else I can do?

bsima commented:
@dotlambda Anything I can do to help move this along?

worldofpeace reviewed (changes requested):
You don`t need to set any of these meta attributes

```
downloadPage = foo;
priority = 0;
broken = false;
updateWalker = true;
```

`downloadPage` could be useful but we already know where we can download python packages from (pypi.org).

Additionally you've set `platforms` to `all`. This is already set to automatically to https://github.com/NixOS/nixpkgs/blob/1b6a4d3979f13f990617d32ee9ec99b1f02673ec/pkgs/development/interpreters/python/mk-python-derivation.nix#L106
which should be sufficient.

worldofpeace commented:
@bsima 
Could you also `fixup` these additional commits with an interactive rebase?
It would make reviewing commit by commit possible.

bsima commented:
worldofpeace <notifications@github.com> writes:

> @bsima 
> Could you also `fixup` these additional commits with an interactive rebase?
> It would make reviewing commit by commit possible.

Yes, I rebased to make things more clear, and updated the meta
info. Thanks for the feedback, and let me know if there is anything else
I can do.

worldofpeace commented on pkgs/development/python-modules/flexmock/default.nix:
```diff
@@ -0,0 +1,29 @@
+  };
+
+  checkInputs = [ nose pytest ];
+
```
Noticed the tests weren't being ran.

This should fix that:
```suggestion
  checkPhase = ''
    pytest tests/
  '';
```

worldofpeace commented on pkgs/development/python-modules/flexmock/default.nix:
```diff
@@ -0,0 +1,29 @@
+
+  meta = with lib; {
+    homepage = https://flexmock.readthedocs.io/en/latest/;
+    description = "flexmock is a testing library for Python.";
```
`description` should not have the name of said software in it or a period at the end.
> https://nixos.org/nixpkgs/manual/#sec-standard-meta-attributes
> Don’t include a period at the end. Don’t include newline characters. Capitalise the first character. For brevity, don’t repeat the name of package — just describe what it does. 

```suggestion
    description = "A testing library for Python that makes it easy to create mocks, stubs, and fakes";
```

worldofpeace commented on pkgs/development/python-modules/flexmock/default.nix:
```diff
@@ -0,0 +1,29 @@
+  meta = with lib; {
+    homepage = https://flexmock.readthedocs.io/en/latest/;
+    description = "flexmock is a testing library for Python.";
+    longDescription = ''
```
This description isn't long so we don't need it.

worldofpeace commented on pkgs/development/python-modules/pgpy/default.nix:
```diff
@@ -0,0 +1,32 @@
+}:
+
+buildPythonPackage rec {
+  pname = "PGPy";
```
I think we're normalizing `pname` now so this should be
```suggestion
  pname = "pgpy";
```

worldofpeace commented on pkgs/development/python-modules/pgpy/default.nix:
```diff
@@ -0,0 +1,32 @@
+  version = "0.4.3";
+
+  src = fetchPypi {
+    inherit pname version;
```
Changing the `pname` will most likely make this not work here.
So we might need:
```suggestion
    pname = "PGPy";
    inherit version;
```

worldofpeace commented on pkgs/development/python-modules/pgpy/default.nix:
```diff
@@ -0,0 +1,32 @@
+
+  meta = with lib; {
+    homepage = https://github.com/SecurityInnovation/PGPy;
+    description = "Pretty Good Privacy for Python 2 and 3.";
```
```suggestion
    description = "Pretty Good Privacy for Python 2 and 3";
```

worldofpeace commented on pkgs/development/python-modules/pgpy/default.nix:
```diff
@@ -0,0 +1,32 @@
+    sha256 = "04412dddd6882ac0c5d5daf4326c28d481421851a68e25e7ac8e06cc9dc2b902";
+  };
+
+  propogatedBuildInputs = [ cryptography pyasn1 six singledispatch ];
```
See https://github.com/SecurityInnovation/PGPy/blob/f1c3d68e32c334f5aa14c34580925e97f17f4fde/setup.py#L36
> only depend on enum34 if Python is older than 3.4

We can constrain the addition of that with
```suggestion
  propogatedBuildInputs = [ cryptography pyasn1 six singledispatch ] ++ lib.optional (pythonOlder "3.4") enum34;
```

worldofpeace commented on pkgs/development/python-modules/py-gfm/default.nix:
```diff
@@ -0,0 +1,29 @@
+    sha256 = "ef6750c579d26651cfd23968258b604228fd71b2a4e1f71dea3bea289e01377e";
+  };
+
+  buildInputs = [ markdown ];
```
```suggestion
  propagatedBuildInputs = [ markdown ];
```

worldofpeace commented on pkgs/development/python-modules/sqlalchemy-utils/default.nix:
```diff
@@ -0,0 +1,34 @@
+}:
+
+buildPythonPackage rec {
+  pname = "SQLAlchemy-Utils";
```
I think we're normalizing pname now so this should be
```suggestion
  pname = "sqlalchemy-utils";
```

worldofpeace commented on pkgs/development/python-modules/sqlalchemy-utils/default.nix:
```diff
@@ -0,0 +1,34 @@
+  version = "0.33.6";
+
+  src = fetchPypi {
+    inherit pname version;
```
Changing the pname will most likely make this not work here.
So we might need:
```suggestion
    pname = "SQLAlchemy-Utils";
    inherit version;
```

worldofpeace commented on pkgs/development/python-modules/sqlalchemy-utils/default.nix:
```diff
@@ -0,0 +1,34 @@
+    sha256 = "45ab41c90bfb8dd676e83179be3088b3f2d64b613e3b590187163dd941c22d4c";
+  };
+  checkInputs = [ pytest flexmock mock python-dateutil pytz ];
+  buildInputs = [ six sqlalchemy ];
```
```suggestion
  propagatedBuildInputs = [ six sqlalchemy ];
```

worldofpeace commented on pkgs/development/python-modules/sqlalchemy-utils/default.nix:
```diff
@@ -0,0 +1,34 @@
+
+  meta = with lib; {
+    homepage = https://github.com/kvesteri/sqlalchemy-utils;
+    description = "Various utility functions and datatypes for SQLAlchemy.";
```
```suggestion
    description = "Various utility functions and datatypes for SQLAlchemy";
```

worldofpeace commented on pkgs/development/python-modules/sshpubkeys/default.nix:
```diff
@@ -0,0 +1,33 @@
+    sha256 = "b388399caeeccdc145f06fd0d2665eeecc545385c60b55c282a15a022215af80";
+  };
+
+  buildInputs = [ cryptography ecdsa ];
```
```suggestion
  propagatedBuildInputs = [ cryptography ecdsa ];
```

worldofpeace commented on pkgs/development/python-modules/sshpubkeys/default.nix:
```diff
@@ -0,0 +1,33 @@
+  buildInputs = [ cryptography ecdsa ];
+
+  # For some reason nix can't find the "tests" module.
+  doCheck = false;
```
Seems somethings not right with the pypi tarball.
Everything works fine when I use the github repo.

```nix
src = fetchFromGitHub {
    owner = "ojarva";
    repo = "python-sshpubkeys";
    rev = "v${version}";
    sha256 = "0000000000000000000000000000000000000000000000000000"; # change this
};
```

worldofpeace replied:
Please also add a comment above it explaining why we have to use github.

worldofpeace reviewed (changes requested):
Additionally tests need aren't running or are missing `checkInputs` for:
 - py-gfm
- pgpy

bsima commented:
Closing in lieu of #54425 - much more progress and better code in that one :)
